File: frontend/services/agentRunService.ts
'use client'

import { apiGet, apiPost } from '@/lib/api-client'
import type {
  AgentRun,
  CreateAgentRunRequest,
  Execution,
  ExecutionEvent,
  ExecutionEventsPage,
  TriggerMedium,
  RunPurpose,
} from '@/types/agent-run'

export const agentRunService = {
  list: async (params: {
    release_id?: string
    task_id?: string
    agent_id?: string
    trigger_medium?: TriggerMedium
    run_purpose?: RunPurpose
    status?: string
  }): Promise<AgentRun[]> => {
    const searchParams = new URLSearchParams()
    if (params.release_id) searchParams.set('release_id', params.release_id)
    if (params.task_id) searchParams.set('task_id', params.task_id)
    if (params.agent_id) searchParams.set('agent_id', params.agent_id)
    if (params.trigger_medium) searchParams.set('trigger_medium', params.trigger_medium)
    if (params.run_purpose) searchParams.set('run_purpose', params.run_purpose)
    if (params.status) searchParams.set('status', params.status)
    const res = await apiGet<AgentRun[]>(`runs?${searchParams}`)
    return res ?? []
  },

  get: async (runId: string): Promise<AgentRun> => {
    return apiGet<AgentRun>(`runs/${runId}`)
  },

  create: async (data: CreateAgentRunRequest): Promise<AgentRun> => {
    return apiPost<AgentRun>('runs', data)
  },

  cancel: async (runId: string): Promise<AgentRun> => {
    return apiPost<AgentRun>(`runs/${runId}/cancel`, {})
  },

  retry: async (runId: string): Promise<AgentRun> => {
    return apiPost<AgentRun>(`runs/${runId}/retry`, {})
  },

  listExecutions: async (runId: string): Promise<Execution[]> => {
    const res = await apiGet<Execution[]>(`executions?run_id=${runId}`)
    return res ?? []
  },

  getExecution: async (executionId: string): Promise<Execution> => {
    return apiGet<Execution>(`executions/${executionId}`)
  },

  listExecutionEvents: async (executionId: string): Promise<ExecutionEvent[]> => {
    const res = await apiGet<ExecutionEventsPage>(`executions/${executionId}/events`)
    return res?.events ?? []
  },

  sendMessage: async (executionId: string, message: string): Promise<void> => {
    await apiPost(`executions/${executionId}/message`, { message })
  },
}

File: backend/app/joysafeter_domain/services/agent_run_service.py
"""
AgentRunService — read-only queries for AgentRun entities.

All mutations (create / cancel / retry) go through ExecutionOrchestrator
which publishes events through the EventBus.
"""

from __future__ import annotations

import uuid
from typing import List, Optional

from sqlalchemy.ext.asyncio import AsyncSession

from app.joysafeter_shared.common.app_errors import InvalidRequestError, NotFoundError
from app.joysafeter_domain.models.agent_run import AgentRun
from app.joysafeter_domain.repositories.agent_run import AgentRunRepository


class AgentRunService:
    """Read-only queries for AgentRun entities."""

    def __init__(self, db: AsyncSession):
        self.db = db
        self.run_repo = AgentRunRepository(db)

    async def list_runs(
        self,
        project_id: Optional[str] = None,
        release_id: Optional[uuid.UUID] = None,
        task_id: Optional[uuid.UUID] = None,
        agent_id: Optional[uuid.UUID] = None,
        trigger_medium: Optional[str] = None,
        run_purpose: Optional[str] = None,
        status: Optional[str] = None,
    ) -> List[AgentRun]:
        """List runs filtered by parameters."""
        if agent_id:
            if not project_id:
                raise InvalidRequestError(
                    "project_id is required when filtering by agent_id",
                    code="AGENT_RUN_PROJECT_REQUIRED",
                    data={"filter": "agent_id"},
                )
            return await self.run_repo.find_by_agent_and_project(
                agent_id=agent_id,
                project_id=project_id,
                trigger_medium=trigger_medium,
                run_purpose=run_purpose,
                status=status,
            )
        elif task_id:
            return await self.run_repo.list_by_task_and_project(task_id, project_id)
        elif release_id:
            return await self.run_repo.list_by_release_and_project(release_id, project_id)
        elif project_id:
            return await self.run_repo.list_by_project(project_id)
        else:
            raise InvalidRequestError(
                "Must provide project_id, release_id, task_id, or agent_id",
                code="AGENT_RUN_FILTER_REQUIRED",
            )

    async def get_run(self, run_id: uuid.UUID) -> AgentRun:
        """Get a run by ID."""
        run = await self.run_repo.get(run_id)
        if not run:
            raise NotFoundError("Agent run not found", code="AGENT_RUN_NOT_FOUND", data={"run_id": str(run_id)})
        return run
